linalg/hypre.cpp

```diff
@@ -319,6 +609,66 @@ void HypreParMatrix::CopyCSR_J(hypre_CSRMatrix *hypre_csr, int *J)
    }
 }
 
+// static method
+signed char HypreParMatrix::HypreCsrToMem(hypre_CSRMatrix *h_mat,
+                                          MemoryType h_mat_mt,
+                                          bool own_ija,
+                                          MemoryIJData &mem)
+{
+   const int nr1 = internal::to_int(h_mat->num_rows) + 1;
+   const int nnz = internal::to_int(h_mat->num_nonzeros);
+   mem.I.Wrap(h_mat->i, nr1, h_mat_mt, own_ija);
+   mem.J.Wrap(h_mat->j, nnz, h_mat_mt, own_ija);
+   mem.data.Wrap(h_mat->data, nnz, h_mat_mt, own_ija);
+   const MemoryClass hypre_mc = GetHypreMemoryClass();
+   if (!CanShallowCopy(mem.I, hypre_mc))
+   {
+      const MemoryType hypre_mt = GetHypreMemoryType();
+      MemoryIJData h_mem;
+      h_mem.I.New(nr1, hypre_mt);
+      h_mem.I.CopyFrom(mem.I, nr1);
+      mem.I.Delete();
+      h_mem.J.New(nnz, hypre_mt);
+      h_mem.J.CopyFrom(mem.J, nnz);
+      mem.J.Delete();
+      h_mem.data.New(nnz, hypre_mt);
+      h_mem.data.CopyFrom(mem.data, nnz);
+      mem.data.Delete();
+      mem = h_mem;
+      if (!own_ija)
+      {
+         // FIXME: Even if own_ija == false, it does not necessarily mean we
+         // need to delete h_mat->{i,j,data} even if h_mat->owns_data == true.
```

Comment:
Should this be addressed?

Reply:
It does not seem to be an issue at the moment, so we can leave this for later.